File: _a_arm/matlab/hw11/armSim.m
armParamHW11;  % load parameters

% instantiate arm, controller, and reference input classes 
addpath('../hw8'); arm = armDynamics(P);  
controller = armController(P);  
addpath('../hw_a'); reference = signalGenerator(30*pi/180, 0.05);  
addpath('../hw_a'); disturbance = signalGenerator(0.25, 0.0);

% instantiate the data plots and animation
addpath('../hw_a'); dataPlot = dataPlotter(P);
addpath('../hw_a'); animation = armAnimation(P);

% main simulation loop
t = P.t_start;  % time starts at t_start
y = arm.h();
while t < P.t_end  
    % Propagate dynamics in between plot samples
    t_next_plot = t + P.t_plot;
    while t < t_next_plot % updates control and dynamics at faster simulation rate
        r = reference.square(t);
        d = disturbance.step(t);
        n = 0;  % noise
        u = controller.update(r, y+n);  % Calculate the control value
        y = arm.update(u+d);  % Propagate the dynamics
        t = t + P.Ts; % advance time by Ts
    end
    % update animation and data plots
    animation.update(arm.state);
    dataPlot.update(t, r, arm.state, u);
end

File: _A_arm/opt/armSim.py
import matplotlib.pyplot as plt
import numpy as np
import armParam as P
from signalGenerator import signalGenerator
from armAnimation import armAnimation
from dataPlotter import dataPlotter
from armDynamics import armDynamics

def armSim(controller):
    # instantiate arm, and reference classes
    arm = armDynamics(alpha=0.2)
    reference = signalGenerator(amplitude=30*np.pi/180.0, frequency=0.05)
    disturbance = signalGenerator(amplitude=0.1)
    # instantiate the simulation plots and animation
    dataPlot = dataPlotter()
    animation = armAnimation()
    # simulation starts at time t_start
    t = P.t_start  
    # output of system at start of simulation
    y = arm.h()  
    #-----------------------------------
    # main simulation loop  
    while t < P.t_end:  
        # plot is updated every P.t_plot second
        t_next_plot = t + P.t_plot
        # updates control and dynamics at faster simulation rate P.Ts
        while t < t_next_plot: 
            r = reference.square(t)  # reference input
            d = disturbance.step(t)  # input disturbance
            n = 0.0  #noise.random(t)  # sensor noise
            u = controller.update(r, y + n)  # update controller
            y = arm.update(u + d)  # propagate system
            t = t + P.Ts  # advance time by Ts
        # update animation and data plots
        animation.update(arm.state)
        dataPlot.update(t, r, arm.state, u)
        # the pause causes the figure to display during simulation
        plt.pause(0.0001)  
    # Keeps the program from closing until the user presses a button.
    print('Press key to close')
    plt.waitforbuttonpress()
    plt.close()
